runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/BeamFnMapTaskExecutor.java

```diff
@@ -351,7 +398,115 @@ void updateProgress() {
       }
     }
 
-    private void updateMetrics(BeamFnApi.Metrics metrics) {
+    // Keeping this as static class for this iteration. Will extract to separate file and generalize
+    // when more counter types are added.
+    // todomigryz: define counter transformer factory
+    // that can provide respective counter transformer for different type of counters.
+    // (ie RowCountCounterTranformer, MSecCounterTransformer, UserCounterTransformer, etc)
+    private static class MonitoringInfoToCounterUpdateTransformer {
+
+      private final Map<String, DataflowStepContext> transformIdMapping;
+
+      public MonitoringInfoToCounterUpdateTransformer(
+          final Map<String, DataflowStepContext> transformIdMapping) {
+        this.transformIdMapping = transformIdMapping;
+      }
+
+      // todo: search code for "beam:metrics"... and replace them with relevant enums from
+      // proto after rebasing above https://github.com/apache/beam/pull/6799 that
+      // introduces relevant proto entries.
+      final String BEAM_METRICS_USER_PREFIX = "beam:metric:user";
+
+      private CounterUpdate monitoringInfoToCounterUpdate(MonitoringInfo monitoringInfo) {
+        long value = monitoringInfo.getMetric().getCounterData().getInt64Value();
+        String urn = monitoringInfo.getUrn();
+
+        String type = monitoringInfo.getType();
+
+        // todomigryz: run MonitoringInfo through validation process.
```

Comment:
Should this be fixed before merging?

Reply:
This requires PR mentioned below to go in.
Technically, I can copy data from another PR, but format of counters is still being defined. It will cause more merging and resolving in both PRs.
I suggest we keep it as is for this PR and generalize/fix it in following PR.